api/incubator/src/main/java/io/opentelemetry/api/incubator/common/ArrayBackedExtendedAttributes.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.api.incubator.common;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.api.internal.ImmutableKeyValuePairs;
+import java.util.ArrayList;
+import java.util.Comparator;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+
+@Immutable
+final class ArrayBackedExtendedAttributes
+    extends ImmutableKeyValuePairs<ExtendedAttributeKey<?>, Object> implements ExtendedAttributes {
+
+  // We only compare the key name, not type, when constructing, to allow deduping keys with the
+  // same name but different type.
+  private static final Comparator<ExtendedAttributeKey<?>> KEY_COMPARATOR_FOR_CONSTRUCTION =
+      Comparator.comparing(ExtendedAttributeKey::getKey);
+
+  static final ExtendedAttributes EMPTY = ExtendedAttributes.builder().build();
+
+  @Nullable private Attributes attributes;
+
+  private ArrayBackedExtendedAttributes(
+      Object[] data, Comparator<ExtendedAttributeKey<?>> keyComparator) {
+    super(data, keyComparator);
+  }
+
+  /**
+   * Only use this constructor if you can guarantee that the data has been de-duped, sorted by key
+   * and contains no null values or null/empty keys.
+   *
+   * @param data the raw data
+   */
+  ArrayBackedExtendedAttributes(Object[] data) {
+    super(data);
+  }
+
+  @Override
+  public ExtendedAttributesBuilder toBuilder() {
+    return new ArrayBackedExtendedAttributesBuilder(new ArrayList<>(data()));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  @Nullable
+  public <T> T get(ExtendedAttributeKey<T> key) {
+    return (T) super.get(key);
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public Attributes asAttributes() {
+    if (attributes == null) {
+      AttributesBuilder builder = Attributes.builder();
+      forEach(
+          (extendedAttributeKey, value) -> {
+            AttributeKey<Object> attributeKey =
+                (AttributeKey<Object>) extendedAttributeKey.asAttributeKey();
+            if (attributeKey != null) {
+              builder.put(attributeKey, value);
+            }
+          });
+      attributes = builder.build();
+    }
```

Comment:
This check-then-set isn't threadsafe, and there's no memory barrier around the `attributes` field. Two threads calling `asAttributes()` on the same instance could receive separate instances...and technically it makes this class mutable. I'm not convinced that it's necessary to fix, however, just want to give it some consideration.

Reply:
`attributes` is just a memoized immutable view of the immutable data represented by this attributes. Similar to caching a hashcode, the race condition between difference threads because all it does is result in possibly re-computing it and small extra memory allocation.

We use this pattern in a variety of places, e.g. [ImmutableKeyValuePairs#hashCode](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/internal/ImmutableKeyValuePairs.java#L249-L259)